orix/tests/quaternion/test_quaternion.py

```diff
@@ -282,3 +283,31 @@ def test_outer_dask_wrong_type_raises(self):
         other = np.random.rand(7, 3)
         with pytest.raises(TypeError, match="Other must be Quaternion or Vector3d"):
             q._outer_dask(other)
+
+    def test_from_align_vectors(self):
+        a = Vector3d([[2, -1, 0], [0, 0, 1]])
+        b = Vector3d([[3, 1, 0], [-1, 3, 0]])
+        ori = Quaternion.from_align_vectors(a, b)
+        assert type(ori) == Quaternion
+        assert np.allclose(
+            ori.data, np.array([[0.65328148, 0.70532785, -0.05012611, -0.27059805]])
+        )
+        assert np.allclose(a.unit.data, (ori * b.unit).data)
+        _, e = Quaternion.from_align_vectors(a, b, return_rmsd=True)
+        assert type(e) == np.float64
+        _, m = Quaternion.from_align_vectors(a, b, return_sensitivity=True)
+        assert type(m) == np.ndarray
+        assert m.shape == (3, 3)
+        out = Quaternion.from_align_vectors(
+            a, b, return_rmsd=True, return_sensitivity=True
+        )
+        assert len(out) == 3
+
+    def test_from_scipy_rotation(self):
+        euler = np.array([15, 32, 41]) * np.pi / 180
+        reference_rot = Orientation.from_euler(euler)
+        scipy_rot = SciPyRotation.from_euler("zxz", euler)  # bunge convention
```

Comment:
A very good test, just shows how horrible Euler angles are! Two changes are important:
1. Using Euler angles, a SciPy rotation is the inverse of an orix rotation, so we have to invert the orix one
2. SciPy accepts both extrinsic rotations (about a fixed reference frame) and intrinsic rotations (about a reference frame moving with the rotations). Rotations in orix assume the latter, hence we have to capitalize rotation axes.

```suggestion
        reference_rot = Rotation.from_euler(euler, direction="crystal2lab")
        scipy_rot = SciPyRotation.from_euler("ZXZ", euler)  # bunge convention
```

One could argue that we should invert rotations within `Quaternion.from_scipy_rotation()` before returning them because of point 2. above. What do you say, @argerlt and @harripj?

Reply:
Actually, this test should be moved to `test_rotation.py`, with `Rotation.from_scipy_rotation()` called instead.

Reply:
Do you have time to consider my point 2. above, @harripj? I'll merge this PR after that is settled, with changes made if we go for the inversion.

Reply:
Hi, apologies for radio silence, I've been writing my candidacy document and had time for little else.
Here are my thoughts on both from_scipy and the unit test:
All of scipy is written using active notation and the vector-scalar quaternion used in computer graphics. All of orix is written in passive using scalar vector quaternions from the Rowenhorst paper.  So, I think it makes sense when writing the conversion to assume the input is given in active and needs to be converted to passive.  to make this clear, there should be a value in the function such as:
`Orientation.from_scipy_rotation(cls, input_interpretation='Active')

Secondly though, I the problem with the tests is related to how each program handles the double cover of quaternions across rotation space. Rotations are part of the Lie Group SO3, but quaternions are part of Lie group SU2, which is related homomorphically, and has two quaternions that related to each unique rotation.  Rowenhorst et al. solves this explicitly by giving rules for casting quaternions to the "Northern hemisphere". (ie, q0 is always >0) . However, scipy does not. 
I think this is intentional: using the double cover of quaternions lets people encode inversion operations to their symmetry rotations. For example, (-0.5.0.5,0.5,0.5) and (0.5.0.5,0.5,0.5)  are quaternions that both represent the same 3-fold spin around <111>, but by saying "if q0 is negative, include inversion" expands the symmetry definition in a handy way. However, scipy was designed by a group of people, not a single person, and I think they only address the problems this caused as they come up. thus, problems such as the ones here.

The solution is to not use quaternions, but instead use rotation matrices. rotation matrices are the ONLY representation in either orix or scipy that is actually isomorphic with SO3 (ie, one-to-one and a continuous smooth manifold). Thus, scipy's quat_to_euler formula correctly handles this double cover ambiguity with it's own internal rules. 

This code demonstrates what I am talking about:
```
# passive bunge representation of an Euler angle triplet
passive_bunge = np.array([53, 14, 11])*np.pi/180
# The equivalent active rotation is reversed and the negative (ie, inverted)
active_bunge = passive_bunge[::-1]*(-1)
print("Passive representation:{}".format(passive_bunge))
print("Active representation:{}".format(active_bunge))


# A vector-scalar active quaternion can be made as thus, using capital letters
# to denote extrinsic rotations:
scipy_quat_from_active = R.from_euler('ZXZ',active_bunge)
# or equivalently, created from the passive bunge as thus:
scipy_quat_from_passive = R.from_euler('ZXZ',passive_bunge[::-1]*(-1))
#A scalar-vector passive quaternion can be made as thus:
orix_quat = Orientation.from_euler(passive_bunge)
print("\nscipy quaternion from passive bunge:{}".format(scipy_quat_from_passive.as_quat()))
print("orix quaternion from passive bunge:{}".format(orix_quat.data))


# However, while orix quaternions are uniquely defined,scipy quaternions are not.
# consider this random choice of rotation matrices:
unnorm_rand_rots = np.random.rand(10000,3,3)-0.5
dets = np.linalg.det(unnorm_rand_rots)
unnorm_rand_rots[dets<0] = unnorm_rand_rots[dets<0]*(-1)
passive_rots = unnorm_rand_rots/((np.linalg.det(unnorm_rand_rots).reshape(10000,1,1))**(1/3))
active_rots = np.stack([x.T for x in passive_rots])
# note how when formed into orix quats, the first value is always positive
orix_quats = Orientation.from_matrix(passive_rots)
# but scipy quats obey no such rules
scipy_quats = R.from_matrix(active_rots)
print("\norix minimum value by column:{}".format(np.min(orix_quats.data, axis=0)))
print("scipy minimum value by column:{}".format(np.min(scipy_quats.as_quat(), axis=0)))
# However, both have internally consistent rules for how to convert to
# rotation matrices, the only isomorphic representation of SO3

passive_bunges = (np.random.rand(1000,3) - 0.5)*np.array([np.pi*2,np.pi,np.pi*2])
active_bunges = passive_bunges[...,::-1]*(-1)
scipy_rot = R.from_euler('ZXZ',active_bunges).as_matrix()
orix_rot = Orientation.from_euler(passive_bunges).to_matrix()
print("\nmaximum difference between conventions when converting using")
print("the rotation matrix : {}".format(np.max(scipy_rot-orix_rot)))
```

Reply:
Thank you for giving this some serious thought, @argerlt.

TL;DR: The only change I suggest to the current implementation is to invert rotations returned from `Quaternion.from_scipy_rotation()`.

@pc494 made an argument against exposing the user to considerations of passive/active rotations in https://github.com/pyxem/orix/issues/249#issue-1074368445:

> I think we should make almost all from_euler code (and similar) private so that we can maintain a single internal set of conventions that are consistently applied. As an example, I believe the "lab2crystal" and "crystal2lab" arguments set whether imported data is "active" (static if your transform3d) or "passive" (rotating in transform3d) - this can get extremely confusing very quickly. Basically I don't think we want this exposed to users.

I agree 100%, and I advice against adding a parameter for toggling passive/active rotation interpretation, and aven mentioning it in the docstring. I think the only necessary interpretation is that a transformation is from the lab to the crystal reference frames, and the inverse transformation can be obtained by inverting the rotation. What do you think, @argerlt?

My initial motivation for not inverting rotations internally is because then SciPy's `Rotation.apply(v)` and orix' `Rotation * Vector3d` produce the same vectors. With the implementation in this PR:

```python
>>> import numpy as np
>>> import scipy.spatial.transform.Rotation as SciPyRotation
>>> from orix.quaternion import Rotation 
>>> from orix.vector import Vector3d
>>> q1, q2, q3, q0 = [0, 0, np.sin(np.pi/4), np.cos(np.pi/4)]
>>> rot_sp = SciPyRotation([q1, q2, q3, q0])
>>> rot = Rotation.from_scipy_rotation(rot_sp)
>>> v = [1, 2, 3]
>>> rot * Vector3d(v)
Vector3d (1,)
[[-2.  1.  3.]]
>>> rot_sp.apply(v)
[-2.  1.  3.]
>>> (~rot) * Vector3d(v)
Vector3d (1,)
[[ 2. -1.  3.]]
```

But, I guess with SciPy's interpretation of a rotation as active means we should invert internally, as you say @argerlt. The above example can be used in the docstring to clearly show the different interpretations of rotations in the two packages.

Reply:
> I think we should make almost all from_euler code (and similar) private so that we can maintain a single internal set of conventions that are consistently applied. As an example, I believe the "lab2crystal" and "crystal2lab" arguments set whether imported data is "active" (static if your transform3d) or "passive" (rotating in transform3d) - this can get extremely confusing very quickly. Basically I don't think we want this exposed to users.
> 
> I agree 100%, and I advice against adding a parameter for toggling passive/active rotation interpretation, and even mentioning it in the docstring. I think the only necessary interpretation is that a transformation is from the lab to the crystal reference frames, and the inverse transformation can be obtained by inverting the rotation. What do you think, @argerlt?

Okay, I can be on board with that. I see the wisdom, and as long as edge cases like this are dealt with internally in a self-consistent manner, no need to further obfuscate the problem with Active/Passive terminology.

>But, I guess with SciPy's interpretation of a rotation as active means we should invert internally, as you say @argerlt. The above example can be used in the docstring to clearly show the different interpretations of rotations in the two packages.

YES!! Okay, this is actually a way better way to look at this than my long winded explanation. Scipy is active, Orix is passive, invert the rotation in from_scipy_rotations automatically internally, and avoid active/passive language in the docstrings. 

@anderscmathisen  I'm not positive, but I think you MAY still run into issues in your unittests when checking how scipy and Orix interpret random Bunge triplets. If not, ignore this next part because your implementation is faster. If so though, my function was also overkill; just replace quaterion.py lines 531-533 in the current PR:

```
 b, c, d, a = rotation.as_quat()

return cls([a, b, c, d])
```
with
```
return cls.from_matrix(np.linalg.inv(rotation.as_matrix()))
```

Reply:
I too, definitively agree that introducing the passive vs active conventions onto the user is not a good idea. It very quickly becomes very confusing. I am quite confused about the matter myself. 

With regards to your concerns about issues with the current implementaiton of `from_scipy_rotation()`, @argerlt, it seems to me like you are absolutely correct. I ran the following test, which creates a bunch of random euler angles, and basically checks if the `orix` `from_euler` method produces the same result as `Scipy` `from_euler` combined with `from_scipy_rotation`. And it appears that the current implementation is not working. Please let me know if you spot any errors in the test.

```python
from orix.quaternion import Rotation, Quaternion
from scipy.spatial.transform import Rotation as SciPyRotation
import numpy as np

num = 100000
eulers = np.random.random(num*3)
eulers = eulers.reshape((-1,3))
eulers *= np.array([2*np.pi, np.pi, 2*np.pi])

correct_current = 0
correct_matrix = 0
for euler in eulers:

    reference_rot = Rotation.from_euler(euler, direction="crystal2lab")
    scipy_rot = SciPyRotation.from_euler("ZXZ", euler)  
    
    current_implementation_way = Quaternion.from_scipy_rotation(scipy_rot)

    matrix_way = Quaternion(Rotation.from_matrix(scipy_rot.as_matrix()).data)

    if np.allclose(current_implementation_way.data, reference_rot.data):
        correct_current += 1

    if np.allclose(matrix_way.data, reference_rot.data):
        correct_matrix += 1

print(f"With current implementaiton: {correct_current/num * 100}")
print(f"With matrix implementaiton: {correct_matrix/num * 100}")
```
Output
```
With current implementaiton: 24.804000000000002
With matrix implementaiton: 100.0
```

So from this, the matrix way is definitively the correct way. 

 
> @anderscmathisen I'm not positive, but I think you MAY still run into issues in your unittests when checking how scipy and Orix interpret random Bunge triplets. If not, ignore this next part because your implementation is faster. If so though, my function was also overkill; just replace quaterion.py lines 531-533 in the current PR:
> 
> ```
>  b, c, d, a = rotation.as_quat()
> 
> return cls([a, b, c, d])
> ```
> 
> with
> 
> ```
> return cls.from_matrix(np.linalg.inv(rotation.as_matrix()))
> ```

The suggested solution here will however not work, as the `Quaternion` class currently does not have an implementation of `from_matrix`. The `from_matrix` method can not be "borrowed" from the `Rotation` class either. eg:
 ```python
rot = Rotation.from_matrix(np.linalg.inv(rotation.as_matrix()))
return cls(rot.data)
```


 as importing `Rotation` into `quaternion.py` leads to circular import errors.

 One possibility is to move the `from_matrix` class method of `Rotation` to the `Quaternion` class?

Reply:
> One possibility is to move the from_matrix class method of Rotation to the Quaternion class?

This is  how I would suggest doing it. As part of both #345 and #373 I've suggested we move all the to/from functions to the Quaternion class. I got positive feedback from @hakonanes , but then got busy and never actually followed through with a completed PR, closest I got was [this branch](https://github.com/argerlt/Mart2Aust_Hackathon/tree/to_from_in_quat) which needs to be redone anyway. I bit off more than I could chew in this branch by also trying to eliminate the Neo Eulerian class, which should have just been a separate PR.

I'd say move just from_matrix, then you or I can open an issue to move the other functions  as part of a later PR. Either that, or someone can write a PR to move all of them now, commit it, then rebase this PR from that one. @hakonanes? @harripj @anderscmathisen  thoughts?

Reply:
Just threw together a new PR for this #406.

Reply:
Hm, I do not see the need for orientation matrices. If we invert the SciPy rotation before extracting the quaternion elements in `from_scipy_rotation()`

```python
        b, c, d, a = rotation.inv().as_quat().T
        return cls(np.column_stack([a, b, c, d]))
```

I get this using your example, @anderscmathisen 

```python
import numpy as np
from orix.quaternion import Rotation
from scipy.spatial.transform import Rotation as SciPyRotation


eulers = Rotation.random((100_000,)).to_euler()

rot_sp = SciPyRotation.from_euler("ZXZ", eulers)
rot_direct = Rotation.from_scipy_rotation(rot_sp)
rot_via_matrix = Rotation.from_matrix(rot_sp.as_matrix())

rot_ref = Rotation.from_euler(eulers)  # lab2crystal
print(rot_ref.angle_with(rot_direct).mean())
# Out: 0.0
print(rot_ref.angle_with(rot_via_matrix).mean())
# Out: 1.5730901506065411
print(rot_ref.angle_with(~rot_via_matrix).mean())
# Out: 0.0
```

We see that by creating orix rotations from SciPy rotations via matrices *without inverting the latter*, as we have now decided that we should, we get a misorientation angle of pi / 2. This angle disappears when inverting the orix reference rotations.

Now that we invert the SciPy rotations, we must also flip the interpretation of the `initial` and `other` reference frames in `Quaternion.from_align_vectors()` by switching the a, b vectors:

```python
        vec1 = initial.unit.data
        vec2 = other.unit.data

        out = SciPyRotation.align_vectors(
            vec1, vec2, weights=weights, return_sensitivity=return_sensitivity
        )
```

With these changes, my previous examples run as before. Of course, some tests fail, which must be updated.

Lastly, moving some of the rotation class methods upstream to the `Quaternion` class is a good idea (I hope to review your new PR soon, @anderscmathisen).

Reply:
From your test here, @hakonanes, I agree that the quaternion conversion works. 

The quaternions made from `from_euler` vs `from_scipy_rotation` via Scipys `from_euler` are, however, not identical in the sense that the elements are not equal, as my test below shows. With the matrix way, they are identical.
But since the angle_with always is zero, I guess the quaternion way is equivalent and that is probably fine?   


```python
import numpy as np
from orix.quaternion import Rotation
from scipy.spatial.transform import Rotation as SciPyRotation

num = 100000
eulers = Rotation.random((num,)).to_euler()

rot_sp = SciPyRotation.from_euler("ZXZ", eulers)
rot_direct = Rotation.from_scipy_rotation(rot_sp)
rot_via_matrix = Rotation.from_matrix(rot_sp.as_matrix())
rot_ref = Rotation.from_euler(eulers)  # lab2crystal

print(rot_ref.angle_with(rot_direct).mean())
# Out: 0.0
print(rot_ref.angle_with(rot_via_matrix).mean())
# Out: 1.5730901506065411
print(rot_ref.angle_with(~rot_via_matrix).mean())
# Out: 0.0

#Find the persentage of rotations where a, b, c and d are 
#all equal to the reference.

direct = np.isclose(rot_ref.data, rot_direct.data) 
print(np.count_nonzero(np.all(direct, axis=1)) / num * 100)
# Out: 75.047

matrix = np.isclose(rot_ref.data, rot_via_matrix.data)
print(np.count_nonzero(np.all(matrix, axis=1)) / num * 100)
# Out: 0

matrix_inv = np.isclose(rot_ref.data, (~rot_via_matrix).data)
print(np.count_nonzero(np.all(matrix_inv, axis=1)) / num * 100)
# Out: 100
```

Reply:
I don't understand why not using quaternions work, but have given up trying to. I'm happy with using orientation matrices. Let's finish #406 before continuing here.

Reply:
#406 is in so we can finish this. Please let me know if you need any assistance @anderscmathisen.